File: Project/Java/Moviesion02/src/ca/ryerson/ee/coe848/pl/Moviesion.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ca.ryerson.ee.coe848.pl;

import ca.ryerson.ee.coe848.cmn.Director;
import ca.ryerson.ee.coe848.cmn.Movie;
import ca.ryerson.ee.coe848.dal.MovieFileManager;
import java.io.IOException;
import java.util.Scanner;

/**
 *
 * @author hosseinfani
 */
public class Moviesion {
    private static final String movieFileName = "Movies.txt";
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        Scanner console = new Scanner(System.in);
        String commandList = 
                  "1) Add Movie\n"
                + "2) Print Movie by Id\n"
                + "3) *Print Movie by Name\n"
                + "4) *Print All Movies\n"
                + "5) *Print All Movies by a Director\n"
                + "6) (Bug)Edit Movie by Id\n"
                + "7) Delete Movie by Id\n"
                + "0) Quit\n"
                + "Enter command: ";
        System.out.println(commandList);
        boolean running = true;
        Integer commandId;
        while (running) {
            try {
                try {
                    commandId = Integer.parseInt(console.nextLine());
                } catch (Exception ex) {
                    System.out.println("Command not recognized!");
                    commandId = -1;
                }
                switch (commandId) {

                    case -1:
                        break;

                    case 0:
                        System.out.println("Bye.");
                        running = false;
                        break;
                    case 1:
                        addMovie(console);
                        break;
                    case 2:
                        printMovieById(console);
                        break;
                    case 3:
                        printMovieByName(console);
                        break;
                    case 4:
                        printAllMovies(console);
                        break;
                    case 5:
                        printAllMoviesByDirector(console);
                        break;
                    case 6:
                        editMovieById(console);
                        break;
                    case 7:
                        deleteMovieById(console);
                        break;
                    
                    default:
                        System.out.println("Command not recognized!");
                        break;
                }
                System.out.println(commandList);
                        
            } catch (Exception ex) {
                System.out.println("Command failed!");
                ex.printStackTrace();
                System.out.println(commandList);
            }
        }
        console.close();
        
    }

    private static void addMovie(Scanner console) throws Exception {
        System.out.println("Enter movie title:");
        String movieTitle = console.nextLine();
        Movie newMovie = new Movie();
        newMovie.setTitle(movieTitle);
        
        //We do not know what is the max Id for movies in our database
        //MovieFileManger knows it
        //newMovie.setId(movieId);
        
        System.out.println("Enter director name: ");
        String directorName = console.nextLine();
        Director newDirector = new Director();
        newDirector.setName(directorName);
        
        //Relationship between movie and director
        newMovie.setDirector(newDirector);
        
        //Add to our movie database in file. 
        //MovieFileManger return the Id of the new movie if the addition was successfull
        //Otherwise it throws an exception
        MovieFileManager mfm = new MovieFileManager(movieFileName);
        Integer id = mfm.add(newMovie);
        newMovie.setId(id);
        
        System.out.println(
                String.format("Movie '%s' has been added with Id='%s'", 
                        newMovie.getTitle(), 
                        newMovie.getId()));
        
    }

    private static void printMovieByName(Scanner console) {
        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
    }

    private static void printMovieById(Scanner console) throws IOException, Exception {
        System.out.println("Enter movie id:");
        Integer id;
        try {
            id = Integer.parseInt(console.nextLine());
        } catch (Exception ex) {
            System.out.println("Movie id not recognized!");
            return;
        }
        MovieFileManager mfm = new MovieFileManager(movieFileName);
        Movie m = mfm.getById(id);
        
        if(m != null)
            System.out.println(m.toString());
        else
            System.out.println("Movie not found!");
    }
    
    private static void printAllMovies(Scanner console) {
        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
    }

    private static void printAllMoviesByDirector(Scanner console) {
        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
    }

    private static void editMovieById(Scanner console) throws IOException, Exception {
        System.out.println("Enter movie id:");
        Integer id;
        try {
            id = Integer.parseInt(console.nextLine());
        } catch (Exception ex) {
            System.out.println("Movie id not recognized!");
            return;
        }
        MovieFileManager mfm = new MovieFileManager(movieFileName);
        Movie m = mfm.getById(id);
        if(m != null)
            System.out.println(m.toString());
        else{
            System.out.println("Movie not found!");
            return;
        }
        
        System.out.println("Enter movie's new title:");
        String movieTitle = console.nextLine();
        m.setTitle(movieTitle);
        
        System.out.println("Enter director name: ");
        String directorName = console.nextLine();
        m.getDirector().setName(directorName);
        
        mfm.deleteById(m.getId());
        Integer newId = mfm.add(m);//Wrong! How come?
        
        //Need to write a new function in MovieFileManger to handle edit
        //mfm.edit(m);
        
        Movie editedMovie = mfm.getById(m.getId());
        //editedMovie is null!
        
        editedMovie = mfm.getById(newId);
        System.out.println(editedMovie.toString());        
    }

    private static void deleteMovieById(Scanner console) throws IOException, Exception {
        System.out.println("Enter movie id:");
        Integer id;
        try {
            id = Integer.parseInt(console.nextLine());
        } catch (Exception ex) {
            System.out.println("Movie id not recognized!");
            return;
        }
        MovieFileManager mfm = new MovieFileManager(movieFileName);
        Movie m = mfm.getById(id);
        if(m != null){
            mfm.deleteById(m.getId());
            System.out.println("Movie is deleted!");
        }else{
            System.out.println("Movie not found!");
        }
    }
}

File: Project/C/Moviesion02/ca.uwindsor.cs.comp3150.pl/Moviesion.c
#include <stdio.h>
#include <string.h>

#include "../ca.uwindsor.cs.comp3150.cmn/Director.c"
#include "../ca.uwindsor.cs.comp3150.cmn/Movie.c"
#include "../ca.uwindsor.cs.comp3150.dal/MovieFileManager.c"

#define DBSIZE 100

void addMovie(/*, struct Movie movies[]*/) {
	
	struct Movie newMovie;	
	printf("Enter movie title:\n");	
	scanf("%s", newMovie.title);

	struct Director newDirector;
	printf("Enter director name:\n");
	scanf("%s", newDirector.name);
	
	//Relationship between movie and director
	strcpy(newMovie.director.name, newDirector.name);
	
	//Add to our file movie database
	//movies[movieId] = newMovie;
	//printf("about to insert ...\n");
	int newId = insertMovie(newMovie);
	printf("Movie '%s' has been added with Id='%d'\n", newMovie.title, newId);	
}

void printMovieByName(/*, struct Movie movies[]*/) {
	//find the movie by its name and print out its full info.
}

int printMovieById(/*, struct Movie movies[]*/) {

	printf("Enter movie id:\n");
	int id;
	scanf("%d", &id);
	/*for(int i=0; i < DBSIZE; i = i + 1){
		if(movies[i].id == id){
			printf("Title: %s\nDirector: %s\n", movies[i].title, movies[i].director.name);
			return id;
		}
	}*/
	struct Movie m = getMovieById(id);
	if(m.id > 0)
		printf("Title: %s\nDirector: %s\n", m.title, m.director.name);
	else
		printf("Movie not found!\n");
	return -1;
}

void printAllMoviesByDirector(/*struct Movie movies[]*/) {
	printf("Enter director name: \n");
	char directorName[100];
	scanf("%s", directorName);
	/*for(int i=0; i < DBSIZE; i = i + 1
		if(movies[i].id > 0 && strcmp(movies[i].director.name, directorName) == 0){
			printf("Movie title: %s\n", movies[i].title);
		}                
	}*/
	
	struct Movie* directorMovies = getMoviesByDirector(directorName);
	int i = 0;
	while(directorMovies[i].id > 0){
		printf("Movie title: %s\n", directorMovies[i].title);
		i = i + 1;
	}
		                
}

void printAllMovies(/*struct Movie movies[]*/) {
	//loop over all movies and print out the info.
}

void editMovieById(/*struct Movie movies[]*/) {
	
	//nightmare :(
}

void deleteMovieById(/*struct Movie movies[]*/) {
	//nightmare :(
}

void main() {

	//Our in-memory movie database
	//struct Movie movies[DBSIZE];
		
	char* commandList = 
"1) Add Movie\n\
2) Print Movie by Id\n\
3) *Print Movie by Title\n\
4) *Print All Movies\n\
5) Print All Movies of a Director\n\
6) *Edit Movie by Id\n\
7) *Delete Movie by Id\n\
0) Quit\n\
Enter command: ";
	int running = 1;
	int commandId;
	while (running) {		
		printf(commandList);
		scanf("%d", &commandId);
		switch (commandId) {
			case -1:
				break;
			case 0:
				printf("Bye.\n");
				running = 0;
				break;
			case 1:
				addMovie();
				break;
			case 2:
				printMovieById();
				break;
			case 3:
				printMovieByName();
				break;
			case 4:
				printAllMovies();
				break;
			case 5:
				printAllMoviesByDirector();
				break;
			case 6:
				editMovieById();
				break;
			case 7:
				deleteMovieById();
				break;
			default:
				printf("Command not recognized!");
				break;		
		}
		printf("\n\n");
	
	}
}

